### PR TITLE
Transaction viewer

### PR DESCRIPTION
**Pull request by taoalpha**
image

### CODE REVIEW

**Comment by taoalpha** [comment]
The viewer tab will be only available to harmony networks.

**Comment by polymorpher** [comment]
Thanks. Set to be reviewed this week

**Comment by taoalpha** [comment]
with amount info retrieved from rosetta api:
image

File: code/client/src/pages/Show/TransactionViewer.jsx [hunk 1]
```diff
@@ -0,0 +1,241 @@
...
+  CreateValidator: 'CreateValidator', // This operation creates a validator on the Harmony network for the sender's account.
+  EditValidator: 'EditValidator', // This operation edits the sender's validator information or election status (i.e: validator name and/or being eligible for election).
+  Delegate: 'Delegate', // This operation delegates (or re-delegates from pending undelegations) native ONE tokens from the sender's account to a validator's account.
+  Undelegate: 'Undelegate', // This operation removes delegation from a validator. Note that this operation does not immediately unlock the sender's tokens.
+  CollectRewards: 'CollectRewards', // This operation credits all the staking rewards to the sender's account.
+}
+
+const PAGE_SIZE = 10
+
+function formatOneValue (val) {
```
Comment (polymorpher) [inline code/client/src/pages/Show/TransactionViewer.jsx:34]: ONEUtil.toOne should do the same - also BigInt would not be enough (only good until 2^53)

File: code/lib/api/index.js [hunk 2]
```diff
@@ -807,10 +807,64 @@ const api = {
     }
   },
   explorer: {
-    decodeMethod: (hash) => {
-      const { data } = axios.get(`https://explorer-v2-api.hmny.io/v0/signature/hash/${hash.slice(10)}`)
+    getApi: () => {
+      return activeNetwork === 'harmony-testnet' ? 'https://rpc.s0.b.hmny.io' : 'https://rpc.s0.t.hmny.io'
```
Comment (polymorpher) [inline code/lib/api/index.js:811]: need to check it is not local either. Also this should be in config / overridable by env

Comment (polymorpher) [reply code/lib/api/index.js:811]: Also we have a more dedicated RPC for wallet queries

**Comment by polymorpher** [comment]
We can make the changes later. I am testing locally, and if it works, I will merge it first

**Comment by polymorpher** [comment]
image

**Comment by polymorpher** [comment]
not showing my receive/send transactions

image

**Comment by polymorpher** [comment]
bunch of 500 from rosetta api. Also not showing my send transactions correctly

image
image

**Comment by polymorpher** [comment]
Clicking "Load More" generates one more tab in pagination selector, but doesn't automatically loads more to the table

image

**Comment by polymorpher** [comment]
Also "Load More" doesn't quite work on some of my wallets

File: code/client/src/pages/Show/TransactionViewer.jsx [hunk 3]
```diff
@@ -0,0 +1,241 @@
...
+      meaningfulOp = op
+    }
+  }
+  let opType
+  let amount
+  if (tx.operations?.length) {
+    opType = tx.operations.find(op => op.type && op.type !== OperationType.Gas)?.type
+  }
+  if (!meaningfulOp) {
+    opType = tx.to === address ? 'Receive' : tx.from === address ? 'Send' : 'Unknown'
```
Comment (polymorpher) [inline code/client/src/pages/Show/TransactionViewer.jsx:56]: 1wallet addresses are always `to`, and `from` is always the relayer's address. So this way of parsing is incorrect

**Comment by polymorpher** [comment]
I think the `block state not found for block xxxx` errors are limitations of the rosetta api - the hosting node of the api might not have all the block states available. Let's use the receipt-parsing method (i.e. parsing logs). See more at https://gist.github.com/cylim/d73f890cd856743c706b91bb37ac5890 , also see https://github.com/polymorpher/one-wallet/blob/c5c2ce7154aca932c260992c5e22d5290d2b3d4e/code/lib/api/flow.js#L298

File: code/lib/api/index.js [hunk 4]
```diff
@@ -819,11 +822,91 @@ const api = {
     }
   },
   explorer: {
-    decodeMethod: (hash) => {
-      const { data } = axios.get(`https://explorer-v2-api.hmny.io/v0/signature/hash/${hash.slice(10)}`)
+    getApi: () => {
```
Comment (polymorpher) [inline code/lib/api/index.js:825]: deprecated by rpcBase in `staking` branch. Will refactor after merging this PR.

File: code/lib/api/index.js [hunk 5]
```diff
@@ -819,11 +822,91 @@ const api = {
...
+
+    decodeMethod: async (hash) => {
+      const { data } = await axios.get(`https://explorer-v2-api.hmny.io/v0/signature/hash/${hash.slice(10)}`)
       return data || []
+    },
+
+    getTransactionHistory: async (address, pageSize = 50, pageIndex = 0) => {
+      const { data } = await axios.post(api.explorer.getApi(), {
+        jsonrpc: '2.0',
+        method: 'hmyv2_getTransactionsHistory',
```
Comment (polymorpher) [inline code/lib/api/index.js:850]: Let's use `eth_` since it's more standard.

File: code/lib/api/index.js [hunk 6]
```diff
@@ -819,11 +822,91 @@ const api = {
...
+        ],
+        id: 1
+      })
+      return data?.result?.transactions || []
+    },
+
+    getTransaction: async (tx) => {
+      const { data } = await axios.post(api.explorer.getApi(), {
+        jsonrpc: '2.0',
+        method: 'hmyv2_getTransactionReceipt',
```
Comment (polymorpher) [inline code/lib/api/index.js:869]: `eth_`

File: code/lib/api/index.js [hunk 7]
```diff
@@ -819,11 +822,91 @@ const api = {
...
+        params: [tx.hash],
+        id: 1
+      })
+
+      return { ...tx, ...parseLog(data.result) }
+    },
+  }
+}
+
+function parseLog (tx) {
```
Comment (polymorpher) [inline code/lib/api/index.js:879]: This should be in a different file, probably util. Will refactor later

File: code/lib/api/index.js [hunk 8]
```diff
@@ -819,11 +822,91 @@ const api = {
...
+          tx.errors.push(eventDetail?.message)
+        }
+        log.eventDataMap[eventName] = ONEUtil.abi.decodeParameters(EventParamsMap[eventName].decodeParams, log.data)
+
+        // Use the first event encountered as the display event for the transaction.
+        if (!tx.displayEvent) {
+          tx.displayEvent = eventName
+          const amountParamIndex = EventParamsMap[eventName].amountParamIndex
+          if (amountParamIndex > -1) {
+            const bi = BigInt(log.eventDataMap[eventName][amountParamIndex]) / BigInt(10 ** 14)
```
Comment (polymorpher) [inline code/lib/api/index.js:897]: Should use BN

File: code/lib/api/index.js [hunk 9]
```diff
@@ -819,11 +822,91 @@ const api = {
...
+            pageIndex
+          }
+        ],
+        id: 1
+      })
+      return data?.result?.transactions || []
+    },
+
+    getTransaction: async (tx) => {
+      const { data } = await axios.post(api.explorer.getApi(), {
```
Comment (polymorpher) [inline code/lib/api/index.js:867]: deprecated by rpcBase in `staking` branch. Will refactor after merging this PR.

**Comment by polymorpher** [comment]
I am making a few changes.

1. removing queries related to backlink addresses. I think it is clearer to show transactions related to only the current address being viewed. We can add an option to show transactions from all addresses later
2. remove pagination and use "load more" only. The reason is that we don't know the total number of transactions on the account. We can add a jump-to-date feature later if that's needed
3. show multiple events per transaction, and decode some methods when the signature is known (e.g. commit). Show errors in different style

**Comment by polymorpher** [comment]
image

File: code/client/src/pages/Show/TransactionViewer.jsx [hunk 10]
```diff
@@ -0,0 +1,238 @@
...
+              </Button>
+              <Button onClick={() => clearFilters()} size='small' style={{ width: 90 }}>
+                Reset
+              </Button>
+            </Space>
+          </div>
+        )
+      },
+      filterIcon (filtered) { return <SearchOutlined style={{ color: filtered ? '#1890ff' : undefined }} /> },
+      onFilter (value, record) {
```
Comment (polymorpher) [inline code/client/src/pages/Show/TransactionViewer.jsx:129]: Need to test this a bit later and fix if needed

**Comment by polymorpher** [comment]
TODO: use web3 contract events lookup to support `eth-` networks

**Comment by polymorpher** [comment]
TODO: fix all onRevealSuccess invocations that are not passing the second argument (events)

File: code/lib/api/index.js [hunk 11]
```diff
@@ -889,6 +888,37 @@ const api = {
         totalSupply,
       }
     },
+  },
+  rpc: {
+    getTransactionHistory: async ({ address, pageSize = 50, pageIndex = 0, fullTx = false }) => {
+      const { data } = await rpcBase.post('', {
+        jsonrpc: '2.0',
+        method: 'eth_getTransactionsHistory', // eth_ method is non-standard, but we still want to use it because it returns normalized addresses and transaction hashes
```
Comment (polymorpher) [inline code/lib/api/index.js:896]: Since this is non-standard, we should replace functions relying on this to using web3 contract events instead

**Comment by polymorpher** [comment]
TODO: add some replacers that support raw amount, ERC20 resolution, ERC721/1155 resolutions

**Review by polymorpher — APPROVED** [review]

**Pull request merged**